Improve parsing and retrieve additional data in REST url-details endpoint

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

This PR is part of the effort to achieve the vision outlined in https://github.com/WordPress/gutenberg/issues/31466. Namely the ability to show meta data about a remote URL when inserting a hyperlink using the block editor.

Following input from @hellofromtonya in https://github.com/WordPress/gutenberg/pull/28791#issuecomment-825819648 this PR takes a regex based route to building on the original implementation from https://github.com/WordPress/gutenberg/pull/18042 to add a better mechanism for parsing the information from the remote URL's response body HTML.

This helps to address feedback such as https://github.com/WordPress/gutenberg/issues/27762 and also sets things up nicely for parsing additional detail from the remote URL (eg: favicon, Open Graph meta...etc).

This PR aims to:

- [x] Parse the site **icon** from the HTML in order to retrieve a usable site icon.
- [x] Parse a representative **image** from the HTML to use as a preview.
- [x] Parse a **description** from the HTML to serve as a longer description.

It should also:

- [x] Improve the means by which we parse regex - for example look only in the `<head>` tag.
- [x] Improve parsing of a **title** to account for edge cases where an attribute is added to the `<title>` tag (eg: `<title data-rh="true">BBC - Home</title>`).

Note [the Open Graph spec](https://ogp.me/).

Closes https://github.com/WordPress/gutenberg/issues/27762

## Automated Testing Instructions

Run the test suite with 

```
wp-env run phpunit 'phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist --verbose --filter WP_REST_URL_Details_Controller_Test'
```

The tests themselves reside in `phpunit/class-wp-rest-url-details-controller-test.php`.

## Manual Testing Instructions

* Check out this PR.
* Run `npm run wp-env start` to boot the local testing env.
* Open the Gutenberg local testing env at `http://localhost:8889/wp-admin/` and login.
* Create a new Post.
* Choose one of the following options:

### Option 1

* Open devtools and in the `console` enter:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=https://wordpress.org' } ).then( data => {
    console.log( data );
} );
```

...also try some non-English websites:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=http://www.baidu.com' } ).then( data => {
    console.log( data );
} );
```


You should see a valid REST API response containing 
  - the contents of the title tag from `wordpress.org` (ie: `"Blog Tool, Publishing Platform, and CMS \u2014 WordPress.org"`).
  - the favicon URL from WordPress.org

### Option 2
* Open devtools and in the `console` enter `wpApiSettings.nonce`. You should see a valid nonce returned as a string.
* Copy the nonce.
* In your browser goto: `http://localhost:8889/wp-json/__experimental/url-details/?url=https://wordpress.org&_wpnonce=%YOUR_NONCE_HERE%` - be sure to replace the nonce value with that copied from the previous step.
* You should see a valid REST API response containing 
  - the contents of the title tag from `wordpress.org` (ie: `"Blog Tool, Publishing Platform, and CMS \u2014 WordPress.org"`).
  - the favicon URL from WordPress.org

getdave committed:
Add basic regex to grab site icon

getdave committed:
Retrieve meta description

getdave committed:
Ensure cleanup

getdave committed:
Improve title regex to account for possible attributes on title

getdave committed:
Retrieve OG Image

getdave committed:
Fix linting

getdave committed:
Fix tests to assert on array subset

getdave committed:
Enhance fixture data with more edge cases

getdave committed:
Add tests to ensure new properties are captured for icon, description and image.

getdave committed:
Add more specific yet flexible test for title

getdave committed:
Handle relative resource URLs for icon and image

getdave committed:
Use random user agent string to avoid being blocked by certain websites.

getdave committed:
Account for open graph image property variations

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -208,13 +214,93 @@ private function get_remote_url( $url ) {
+	 * @param string $html the HTML from the remote website at URL.
+	 * @return string the meta description contents (maybe empty).
+	 */
+	private function get_description( $html ) {
```
@hellofromtonya How do you feel about the writing to a `tmp` file here simply to make use of `get_meta_tags`?

hellofromtonya replied:
It's a clever way to use native PHP functionality. However, I'm concerned about working with the filesystem: server permissions, leaving behind artifacts, performance, etc. 

For this specific parsing method, it's too much for parsing just the `<meta name="description" ... >` element. Why? The meta element has a clear well-formed signature (via the `name` attribute). Here, regex will be more understandable and performant without the potential filesystem gotchas that some sites will experience.

Side note:
This type of code, i.e. get all meta elements, could be something for pre-parsing the `head` _before_ invoking each specific parser (i.e. before getting the title, icon, description, image, etc.). However, it would need consideration for performance, memory handling, cognitive complexity, etc. I'm noting it here for thought when we start working on tuning the performance of getting the metadata.

getdave replied:
Unfortunately, it seems to only return a very specific subset of `meta` tags. At least when I tested.

Let's switch to regex here then.

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -208,13 +214,92 @@ private function get_remote_url( $url ) {
+
+		// Attempt to convert relative URLs to absolute.
+		if ( ! empty( $image ) ) {
+			$image = \WP_Http::make_absolute_url( $image, $url );
```
@hellofromtonya Some URLs will be relative. Is this a good method to use to convert to absolute?

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -208,13 +214,92 @@ private function get_remote_url( $url ) {
+	 * @return string the OG image (maybe empty).
+	 */
+	private function get_image( $html, $url ) {
+		preg_match( '|<meta.*?property="og:image[:url]*?".*?content="(.*?)".*?\/?>|is', $html, $matches );
```
@hellofromtonya Here we are only interested in

`property="og:image"`

or

`property="og:image:url`

Any other variations should be ignored.

getdave commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -208,13 +214,92 @@ private function get_remote_url( $url ) {
+		$description = ! empty( $meta['description'] ) ? $meta['description'] : '';
+
+		fclose( $temp ); // clean up tmp file.
+		return $description;
```
If a description contains HTML tags (here's [an example](https://www.theguardian.com/world/2021/may/19/eu-to-ask-uk-to-respect-citizens-rights-after-mistreatment-scandals)) in the meta `description` then 

1. Would you advise "cleaning" them?
2. If so...how? 

Often they will be in HTML entity form rather than as literal HTML strings. Presumanly because on the target website they are escaped before they are written to the meta tag.

getdave replied:
How about https://www.php.net/manual/en/function.html-entity-decode.php and then https://developer.wordpress.org/reference/functions/wp_strip_all_tags/.

That would ensure entities are parsed to HTML and then all the HTML is stripped leaving just the content.

Would that work?

hellofromtonya replied:
Depends on the goal of this controller. This controller is:

- parsing the web page and scraping the metadata content out of it
- structuring the metadata for processing downstream

Should decisions be made here in this class or downstream (as it's closer to the render)? Or is it a reasonable expectation that metadata structure will be decoded (i.e. in HTML format) for the downstream processing?

getdave replied:
I guess if HTML is valid in the <meta> description tag then we shouldn't mess with it and instead expect the downstream to do the processing.

hellofromtonya replied:
Agreed. Let's go that way.

getdave committed:
Add unit test for get_title

getdave committed:
Add tests (including some failing) for get_icon

getdave committed:
Fix method invocation to remove unused args

hellofromtonya commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -118,7 +118,10 @@ public function parse_url_details( $request ) {
+				'title'       => $this->get_title( $remote_url_response, $url ),
+				'icon'        => $this->get_icon( $remote_url_response, $url ),
+				'description' => $this->get_description( $remote_url_response, $url ),
+				'image'       => $this->get_image( $remote_url_response, $url ),
```
The work being done here is only in the `<head>`.  To make the string searches faster and more efficient, I'd suggest splitting the HTML to process only the  `<head>`.

How?

Add a new private method to search for the closing `</head>` tag or if missing, the opening `<body>` tag.  Then return the string up to that tag. Then pass this partial HTML to each of these metadata helper methods.

getdave replied:
Agreed. I've added a basic implementation that gets everything before the opening `<body>` tag. That should be good enough for our purposes.

getdave replied:
...although I guess if there wasn't a `<body>` tag that could be an issue.

getdave committed:
Wrap test HTML string in a basic HTML doc.

getdave committed:
Parse the head section and use for comparison

getdave committed:
Fix broken cache test

getdave committed:
Refine wrap method

getdave committed:
Add get_image tests

getdave committed:
Handle relative URLs when target url has a path

Commit:
Improves title and icon parsing for PR 31763 (#32021)

* Title: removes malformed opening tag pattern and adds tests.

* Icon: Allows for different ordering of attribute. Adds happy
and unhappy test data.

* Icon: allow for any order or combination of attributes.

How?

Get the icon link element first.
Then grab its href.

Benefits:
- Not dependent upon the order of attributes
- Allows for optional or custom attributes

* Icon: allows for single, double, or no quotes around attributes.

* Update for WPCS standard.

getdave committed:
Seek head but fallback to body.

Commit:
Improves metadata parsing for PR 31763 (#32067)

* Description: uses regex instead of tmp file.

* Adding test to check for like tag before and after target.

* Description: changes regex strategy.

Why?

Lookahead was not constrained with each element and thus picked up
<meta from one and then if not a match, grabbed the name and content
from another upstream.

The new strategy parses all meta elements with a content
attribute. Then loops through them to find the description element.

Why this order?
The content attribute can contain HTML tags. The > or /> symbol is
matched as the end of the meta element (it's closing symbol). If
this happens, the content is truncated. Boo.

Switching the parsing order solves this problem.

Bonus: allows for pre-parsing of all meta elements. Performance boost.

* Refactors getting meta with content elements for reuse.

* Improves getting <head>..</head> element.

- Isolates to the only the <head>..</head> element by stripping
  all content before the opening tag and ensuring it includes a closing
  </head> tag.
- Performance improvements:
   - Bails out early if no opening tag is found.
   - Uses native string functions instead of regex.

* Image: use same parsing strategy as description.

* Refactor to reuse the process for getting the metadata from the list of meta elements.

* Convert description HTML entities into HTML.

Commit:
Improves PR 31763 for the URL Details Controller (#32162)

* Code standards and consistency.

* Removed unused data provider.

* More formatting and standards.

* Title: converts entities.

* Fixes asserts: removes deprecated array subset, uses assertSame, and makes consistent.

* Fixes method return signatures.

* Remove HTML and convert non-HTML entities.

* Removes type check from set_cache as data will be string type..

TimothyBJacobs commented:
Code seems good to me. I'm -1 on spoofing a user agent though. If a site owner wants to block WordPress from making that API call, that should be their prerogative.

getdave commented:
> I'm -1 on spoofing a user agent though. If a site owner wants to block WordPress from making that API call, that should be their prerogative.

@TimothyBJacobs Fair point. I was finding that a couple of sites seemed to be blocking requests so I assumed if the request looked more like a typical/common user agent then it would be ok.

We can revert as the functionality can still be achieved via a _filter_ if folks want to use my method in their own Plugin/code.

hellofromtonya commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -164,12 +168,12 @@ public function permissions_check() {
 	 *
 	 * @param string $url The website url whose HTML we want to access.
-	 * @return array|WP_Error the HTTP response from the remote URL or error.
+	 * @return array|WP_Error the HTTP response from the remote URL, or an error.
```
```suggestion
	 * @return string|WP_Error The HTTP response from the remote URL, or an error.
```

`wp_remote_retrieve_body` returns a `string`.

hellofromtonya commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -164,12 +168,12 @@ public function permissions_check() {
-
 		$args = array(
 			'limit_response_size' => 150 * KB_IN_BYTES,
+			'user-agent'          => $this->get_random_user_agent(),
```
```suggestion
```

hellofromtonya commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -261,4 +372,205 @@ private function set_cache( $key, $data = '' ) {
+	 *
+	 * @return string the user agent string.
+	 */
+	private function get_random_user_agent() {
```
>We can revert as the functionality can still be achieved via a filter if folks want to use my method in their own Plugin/code.

@getdave did you mean exposing this method for plugins/code? 

If yes, I'd suggest making this method static:

```suggestion
	public static function get_random_user_agent() {
```

Why? 
- Easier to invoke.
- And it's no longer directly associated with the specifics of an object; rather it's now a helper.

That said, a helper like this one seems out-of-place here in this class. When a random user agent is needed, this type of helper might be better served from the REST or HTTP API.

getdave replied:
No I meant you can [filter the args](https://github.com/WordPress/gutenberg/blob/411291d5b7fb26c93b6d777a3f478a97797d0eee/lib/class-wp-rest-url-details-controller.php#L187) in a Plugin or your own code. Agree we should remove the method from the `class`.

hellofromtonya commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -261,4 +372,205 @@ private function set_cache( $key, $data = '' ) {
 
 		return set_transient( $key, $data, $cache_expiration );
 	}
+
+	/**
+	 * Picks a random user agent string from a list of common defaults.
+	 * By default WordPress HTTP functions uses a semi-static string and
+	 * this maybe rejected by many websites.
+	 *
+	 * See: https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/class-http.php#L191.
+	 *
+	 * @return string the user agent string.
+	 */
+	private function get_random_user_agent() {
+		$agents = array(
+			'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246', // Windows 10-based PC using Edge browser.
+			'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/601.3.9 (KHTML, like Gecko) Version/9.0.2 Safari/601.3.9', // Mac OS X-based computer using a Safari browser.
+			'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:15.0) Gecko/20100101 Firefox/15.0.1', // Linux-based PC using a Firefox browser.
+		);
+
+		$chose = rand( 0, count( $agents ) - 1 );
+
+		return $agents[ $chose ];
+	}
```
```suggestion
```

hellofromtonya commented:
@getdave The above 3 suggestions take care of the remaining issues. Once committed, the PR should be ready for merge.

getdave committed:
Update lib/class-wp-rest-url-details-controller.php

Co-authored-by: Tonya Mork <hello@hellofromtonya.com>

getdave committed:
Update lib/class-wp-rest-url-details-controller.php

Co-authored-by: Tonya Mork <hello@hellofromtonya.com>

getdave committed:
Update lib/class-wp-rest-url-details-controller.php

Co-authored-by: Tonya Mork <hello@hellofromtonya.com>

getdave commented:
@hellofromtonya Thanks for these. All committed 🥳 

Let's get a final 👍 from @TimothyBJacobs so I can merge this.

hellofromtonya commented:
Hey @TimothyBJacobs, `user-agent` force code is now removed. Appreciate a code review and, if 👍 , approval.

getdave commented:
@hellofromtonya One for a followup, but I've found [an edge case](http://www.foo.com/) where the icon is provided as a data URI:

eg: 

```html
<link href='data:image/png;base64,iVBORw0KGgo=' rel='icon' type='image/png'>
```

We may need to account for situations such as these.

hellofromtonya commented:
>We may need to account for situations such as these.

@getdave Aw yes, the data URL. ~What do you foresee we need to account for, i.e. special handling needs?~

Most browsers block these for top-level navigation due to security concerns. But it seems okay for a favicon. Testing in Firefox and Chrome, both rendered this example:

```
data:image/x-icon;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQEAYAAABPYyMiAAAABmJLR0T///////8JWPfcAAAACXBIWXMAAABIAAAASABGyWs+AAAAF0lEQVRIx2NgGAWjYBSMglEwCkbBSAcACBAAAeaR9cIAAAAASUVORK5CYII=
```
when directly loading the URL into the browser.

UPDATE:
I see. Need to skip over the relative to absolute URL code. On it.

hellofromtonya commented:
@getdave Addressed the icon data URL in PR #32276 and cross-version PHP tested it https://3v4l.org/aaDc9.

Commit:
Icon: if data url, skip relative-to-absolute conversion (#32276)

hellofromtonya commented on phpunit/class-wp-rest-url-details-controller-test.php:
```diff
@@ -488,37 +427,515 @@ public function test_get_item_schema() {
+			),
+			'with data URL x-icon type'             => array(
+				'<link rel="icon" href="data:image/x-icon;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQEAYAAABPYyMiAAAABmJLR0T///////8JWPfcAAAACXBIWXMAAABIAAAASABGyWs+AAAAF0lEQVRIx2NgGAWjYBSMglEwCkbBSAcACBAAAeaR9cIAAAAASUVORK5CYII=" type="image/x-icon" />',
+				'data:image/x-icon;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQEAYAAABPYyMiAAAABmJLR0T///////8JWPfcAAAACXBIWXMAAABIAAAASABGyWs+AAAAF0lEQVRIx2NgGAWjYBSMglEwCkbBSAcACBAAAeaR9cIAAAAASUVORK5CYII="',
```
```suggestion
				'data:image/x-icon;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQEAYAAABPYyMiAAAABmJLR0T///////8JWPfcAAAACXBIWXMAAABIAAAASABGyWs+AAAAF0lEQVRIx2NgGAWjYBSMglEwCkbBSAcACBAAAeaR9cIAAAAASUVORK5CYII=',
```

@getdave the `"` at the end of the expected string shouldn't be there and is causing the tests to fail

Commit:
Fix failing test due to extra character in expected string.

TimothyBJacobs commented:
Code is looking good to me. I haven't dived deep into the regexes, but the tests look good.

The schema needs to be updated to account for the new data we are returning.

Commit:
Updates schema for new data items.

TimothyBJacobs commented on lib/class-wp-rest-url-details-controller.php:
```diff
@@ -66,8 +66,26 @@ public function get_item_schema() {
+					'readonly'    => true,
+				),
+				'icon'        => array(
+					'description' => __( 'The favicon image link of the <link rel="icon"> element from the URL.', 'gutenberg' ),
```
I think this can get a `format` of `uri` right?

TimothyBJacobs replied:
And same for `image`.

hellofromtonya replied:
You're right @TimothyBJacobs. Updated both in commit https://github.com/WordPress/gutenberg/pull/31763/commits/b013a82031c326bffe0b2676d2f21798b0eb2cfb

TimothyBJacobs replied:
Sorry, to clarify, the `type` is still `string` but there should be a `format` keyword with `uri`.

hellofromtonya replied:
Doh, of course. Sorry about that. Reverted and added `format`  in commit https://github.com/WordPress/gutenberg/pull/31763/commits/0097a7407c4cf3dfec7d94f476b0a6c36844714c

Commit:
Changes icon and image type to uri.

Commit:
Schema: icon & image: reverts type back to string and adds format of uri.
